src/Features/Core/Portable/CodeRefactorings/AbstractRefactoringHelpersService.cs

```diff
@@ -0,0 +1,260 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.CodeRefactorings
+{
+    internal abstract class AbstractRefactoringHelpersService : IRefactoringHelpersService
+    {
+        public async Task<TSyntaxNode> TryGetSelectedNodeAsync<TSyntaxNode>(
+            Document document, TextSpan selection, CancellationToken cancellationToken) where TSyntaxNode : SyntaxNode
+        {
+            return await TryGetSelectedNodeAsync(document, selection, n => n is TSyntaxNode, cancellationToken).ConfigureAwait(false) as TSyntaxNode;
+        }
+
+        /// <summary>
+        /// <para>
+        /// Returns a Node for refactoring given specified selection that passes <paramref name="predicate"/> 
+        /// or null if no such instance exists.
+        /// </para>
+        /// <para>
+        /// A node instance is return if:
+        /// - Selection is zero-width and inside/touching a Token with direct parent passing <paramref name="predicate"/>.
+        /// - Selection is zero-width and touching a Token whose ancestor Node passing <paramref name="predicate"/> ends/starts precisely on current selection.
+        /// - Token whose direct parent passing <paramref name="predicate"/> is selected.
+        /// - Whole node passing <paramref name="predicate"/> is selected.
+        /// </para>
+        /// <para>
+        /// Attempts extracting (and testing with <paramref name="predicate"/> a Node for each Node it consideres (see above).
+        /// By default extracts initializer expressions from declarations and assignments.
+        /// </para>
+        /// <para>
+        /// Note: this function trims all whitespace from both the beginning and the end of given <paramref name="selection"/>.
+        /// The trimmed version is then used to determine relevant <see cref="SyntaxNode"/>. It also handles incomplete selections
+        /// of tokens gracefully.
+        /// </para>
+        /// </summary>
+        protected Task<SyntaxNode> TryGetSelectedNodeAsync(Document document, TextSpan selection, Predicate<SyntaxNode> predicate, CancellationToken cancellationToken) => TryGetSelectedNodeAsync(document, selection, predicate, DefaultNodeExtractor, cancellationToken);
```

Comment:
for something this long, wrap.  A good rule of thumb: i it can't be seen on github without scrolling, it's too long.

IN this case, i thought this was an abstract method at first since i saw no body.